### PR TITLE
feat: Upload image in update event fragment

### LINKED ISSUE #1410: Feature to update event image by uploading a picture
**Describe the feature**
Clicking on the ImageView on the Update Events screen should give the user an option to upload an image.

1. Go to Edit Event.
2. Click on image.
3. Nothing happens.
4. User should get an option to select and upload a picture.

**Screenshots**
Nothing happens on clicking the placeholder image.
screenshot_2018-12-26-23-35-11-535_com eventyay organizer

**Would you like to work on the issue?**
- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by ShridharGoel**
Fixes #1410 

Checklist:

- [ ] I have checked for PMD and check-style issues
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream development branch.

Changes: 

Earlier, nothing was happening on clicking the image which could confuse the user on how to update the picture.
Now, on clicking the image in UpdateEventFragment, user would be given the option to upload a picture.

### CODE REVIEW

**Comment by iamareebjamal** [comment]
Image Upload has to be implemented instead of this

**Comment by ShridharGoel** [comment]
@iamareebjamal But I think the system is configured to receive a URL and image is fetched from that URL and saved. How can we upload directly from the mobile?

**Comment by iamareebjamal** [comment]
https://api.eventyay.com/#upload-image-upload-post

**Comment by ShridharGoel** [comment]
@iamareebjamal Getting HTTP 500 Internal Server Error on this.

**Comment by Masquerade0097** [comment]
@ShridharGoel Please share the request details.

**Comment by ShridharGoel** [comment]
@Masquerade0097 I'm sending a POST request with image URI data as a Field parameter.

**Comment by iamareebjamal** [comment]
Worked perfectly for me. 

```{
    "url": "https://open-event-api-dev.herokuapp.com/static/media/temp/images/d004becc-91cf-4c50-84bb-86d52a4448a8/L2E1T2RWaU/3eaa4ae0-f0ac-4df0-a8cd-dc4cf89e8c0a.png"
}```

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 1]
```diff
@@ -101,9 +110,19 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+            intent.setAction(Intent.ACTION_GET_CONTENT);
+            startActivityForResult(Intent.createChooser(intent, "Select Picture"), IMAGE_CHOOSER_REQUEST_CODE);
+        });
+
         binding.submit.setOnClickListener(view -> {
             if (validator.validate()) {
                 createEventViewModel.updateEvent();
+                if(createEventViewModel.imageUrl!=null) {
+                    binding.form.originalImageUrl.setText(createEventViewModel.imageUrl);
+                }
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: Use LiveData instead

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 2]
```diff
@@ -336,9 +355,33 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                 binding.form.description.setText(description);
                 binding.form.description.setTextColor(Color.BLACK);
             }
+        } else if (requestCode == IMAGE_CHOOSER_REQUEST_CODE && resultCode == RESULT_OK) {
+            Uri selectedImageUri = data.getData();
+            Bitmap bitmap;
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: Join the declaration and assignment

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 3]
```diff
@@ -336,9 +355,33 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+            try {
+                InputStream imageStream = getActivity().getContentResolver().openInputStream(selectedImageUri);
+                bitmap = BitmapFactory.decodeStream(imageStream);
+                String encodedImage = encodeImage(bitmap);
+                ImageData imageData = new ImageData("data:image/gif;base64,"+encodedImage);
+                createEventViewModel.uploadImage(imageData);
+                binding.form.eventOriginalImage.setImageBitmap(bitmap);
+            } catch (FileNotFoundException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: Use Timber and display toast

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 4]
```diff
@@ -336,9 +355,33 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
 
+    private String encodeImage(Bitmap bm)
+    {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        bm.compress(Bitmap.CompressFormat.JPEG,100,baos);
+        byte[] b = baos.toByteArray();
+        String encImage = Base64.encodeToString(b, Base64.DEFAULT);
+
+        return encImage;
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: To Utils

File: app/src/main/java/com/eventyay/organizer/data/event/EventApi.java [hunk 5]
```diff
@@ -30,4 +30,7 @@
 
     @GET("events/{id}/general-statistics")
     Observable<EventStatistics> getEventStatistics(@Path("id") long id);
+
+    @POST("upload/image")
+    Observable<OriginalImageTemporaryUrl> postOriginalImage(@Body ImageData data);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/event/EventApi.java]: Shouldn't be in EvwntApi

File: app/src/main/java/com/eventyay/organizer/data/event/ImageData.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    private String data;
+
+    public ImageData(String data) {
+        this.data = data;
+    }
+
+    public String getData() {
+        return data;
+    }
+}
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/event/ImageData.java]: Use lombok

File: app/src/main/java/com/eventyay/organizer/data/event/OriginalImageTemporaryUrl.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public String getUrl() {
+        return url;
+    }
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
+}
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/event/OriginalImageTemporaryUrl.java]: Use lombok

File: app/src/main/java/com/eventyay/organizer/data/event/OriginalImageTemporaryUrl.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package com.eventyay.organizer.data.event;
+
+import lombok.Data;
+
+@Data
+public class OriginalImageTemporaryUrl {
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/event/OriginalImageTemporaryUrl.java]: Rename to ImageUrl

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java [hunk 9]
```diff
@@ -57,6 +58,8 @@
 
     private Event event = new Event();
 
+    public String imageUrl;
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Use LiveData

File: app/src/main/java/com/eventyay/organizer/utils/Utils.java [hunk 10]
```diff
@@ -71,4 +75,14 @@ public static String getShareableInformation(Event event) {
         return data.toString();
     }
 
+    public static String encodeImage(Bitmap bm)
+    {
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/utils/Utils.java]: Use java style guideline

File: app/src/main/java/com/eventyay/organizer/utils/Utils.java [hunk 11]
```diff
@@ -71,4 +75,14 @@ public static String getShareableInformation(Event event) {
...
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        bm.compress(Bitmap.CompressFormat.JPEG,100,baos);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/utils/Utils.java]: Space after ,

**Comment by iamareebjamal** [comment]
Good job

**Comment by ShridharGoel** [comment]
Thanks :)

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java (hunk 9 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Okay

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 12]
```diff
@@ -101,8 +108,19 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+            intent.setAction(Intent.ACTION_GET_CONTENT);
+            startActivityForResult(Intent.createChooser(intent, "Select Picture"), IMAGE_CHOOSER_REQUEST_CODE);
+        });
+
         binding.submit.setOnClickListener(view -> {
             if (validator.validate()) {
+                if(createEventViewModel.getImageUrlLiveData().getValue()!=null) {
+                    binding.form.originalImageUrl.setText(createEventViewModel.imageUrl.getUrl());
+                    createEventViewModel.getEventLiveData().observe(this, this::setEvent);
+                }
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: This is not how LiveData works

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java [hunk 13]
```diff
@@ -96,6 +101,11 @@ public Event getEvent() {
         return eventMutableLiveData;
     }
 
+    public LiveData<ImageUrl> getImageUrlLiveData() {
+        imageUrlMutableLiveData.setValue(imageUrl);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Remove this line

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java [hunk 14]
```diff
@@ -54,9 +56,12 @@
     private final MutableLiveData<Boolean> progress = new MutableLiveData<>();
     private final MutableLiveData<Void> close = new MutableLiveData<>();
     private final MutableLiveData<Event> eventMutableLiveData = new MutableLiveData<>();
+    private final MutableLiveData<ImageUrl> imageUrlMutableLiveData = new MutableLiveData<>();
 
     private Event event = new Event();
 
+    public ImageUrl imageUrl;
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Remove this

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java [hunk 15]
```diff
@@ -250,4 +260,17 @@ public void updateEvent() {
...
+    public void uploadImage(ImageData imageData) {
+        compositeDisposable.add(
+            eventRepository
+                .uploadEventImage(imageData)
+                .doOnSubscribe(disposable -> progress.setValue(true))
+                .doFinally(() -> progress.setValue(false))
+                .subscribe(uploadedImage -> {
+                    onSuccess.setValue("Image Uploaded Successfully");
+                    imageUrl = uploadedImage;
+                    getImageUrlLiveData();
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Set value here

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 16]
```diff
@@ -101,8 +109,16 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+        binding.form.eventOriginalImageLayout.setOnClickListener(view -> {
+            Intent intent = new Intent();
+            intent.setType("image/*");
+            intent.setAction(Intent.ACTION_GET_CONTENT);
+            startActivityForResult(Intent.createChooser(intent, "Select Picture"), IMAGE_CHOOSER_REQUEST_CODE);
+        });
+
         binding.submit.setOnClickListener(view -> {
             if (validator.validate()) {
+                createEventViewModel.getImageUrlLiveData().observe(this, this::setImageUrl);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: Observe it outside the click listener

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java [hunk 17]
```diff
@@ -336,6 +352,19 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                 binding.form.description.setText(description);
                 binding.form.description.setTextColor(Color.BLACK);
             }
+        } else if (requestCode == IMAGE_CHOOSER_REQUEST_CODE && resultCode == RESULT_OK) {
+            Uri selectedImageUri = data.getData();
+            try {
+                InputStream imageStream = getActivity().getContentResolver().openInputStream(selectedImageUri);
+                Bitmap bitmap = BitmapFactory.decodeStream(imageStream);
+                String encodedImage = Utils.encodeImage(bitmap);
+                ImageData imageData = new ImageData("data:image/gif;base64,"+encodedImage);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: Why data/gif?

File: app/src/main/java/com/eventyay/organizer/utils/Utils.java [hunk 18]
```diff
@@ -71,4 +75,13 @@ public static String getShareableInformation(Event event) {
         return data.toString();
     }
 
+    public static String encodeImage(Bitmap bm) {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        bm.compress(Bitmap.CompressFormat.JPEG,100, baos);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/utils/Utils.java]: Space after comma

File: app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java (hunk 17 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: It's defined like that in the API, isn't it?

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: No, it's an example

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: But when I tried without this, it didn't work

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: You have to add proper meta, not just image/gif always

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/create/UpdateEventFragment.java]: Done

File: app/src/main/java/com/eventyay/organizer/utils/Utils.java [hunk 19]
```diff
@@ -71,4 +78,16 @@ public static String getShareableInformation(Event event) {
         return data.toString();
     }
 
+    public static String encodeImage(Bitmap bm, Uri selectedImageUri, Context context) {
```
[the remark this answers is not in the document]
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/utils/Utils.java]: Done

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java (hunk 15 above)
Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Still not solved

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Remove this line. Why are we getting the livedata here?

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java [hunk 20]
```diff
@@ -250,4 +257,17 @@ public void updateEvent() {
...
+    public void uploadImage(ImageData imageData) {
+        compositeDisposable.add(
+            eventRepository
+                .uploadEventImage(imageData)
+                .doOnSubscribe(disposable -> progress.setValue(true))
+                .doFinally(() -> progress.setValue(false))
+                .subscribe(uploadedImage -> {
+                    onSuccess.setValue("Image Uploaded Successfully");
+                    imageUrlMutableLiveData.setValue(uploadedImage);
+                    getImageUrlLiveData();
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]:
```suggestion
```

**Comment by iamareebjamal** [comment]
Whenever you create a PR, please check allow changes by maintainers. Or else we'll have to go ping pong of request-review cycle for small changes. I wanted to merge the PR so tried to do the change but couldn't

**Comment by ShridharGoel** [comment]
But the option to allow edits from maintainers was already checked

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java (hunk 20 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Sorry, it's removed now.

File: app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java (hunk 15 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/create/CreateEventViewModel.java]: Sorry, it's removed now.

**Pull request merged**